js.go

```diff
@@ -0,0 +1,1175 @@
+	SubscribeSync(subj string, opts ...SubOpt) (*Subscription, error)
+	// Channel versions.
+	ChanSubscribe(subj string, ch chan *Msg, opts ...SubOpt) (*Subscription, error)
+	// QueueSubscribe.
+	QueueSubscribe(subj, queue string, cb MsgHandler, opts ...SubOpt) (*Subscription, error)
+
+	// Management
+	// TODO(dlc) - add more
+	AddStream(cfg *StreamConfig) (*StreamInfo, error)
+	AddConsumer(stream string, cfg *ConsumerConfig) (*ConsumerInfo, error)
+}
+
+// ApiError is included in all API responses if there was an error.
+type ApiError struct {
+	Code        int    `json:"code"`
+	Description string `json:"description,omitempty"`
+}
+
+// ApiResponse is a standard response from the JetStream JSON API
+type ApiResponse struct {
+	Type  string    `json:"type"`
+	Error *ApiError `json:"error,omitempty"`
+}
+
+type AccountInfoResponse struct {
+	ApiResponse
+	*AccountStats
+}
+
+// AccountLimits is for the information about
+type AccountLimits struct {
+	MaxMemory    int64 `json:"max_memory"`
+	MaxStore     int64 `json:"max_storage"`
+	MaxStreams   int   `json:"max_streams"`
+	MaxConsumers int   `json:"max_consumers"`
+}
+
+// AccountStats returns current statistics about the account's JetStream usage.
+type AccountStats struct {
+	Memory  uint64        `json:"memory"`
+	Store   uint64        `json:"storage"`
+	Streams int           `json:"streams"`
+	Limits  AccountLimits `json:"limits"`
+}
+
+const defaultRequestWait = 5 * time.Second
+
+// Internal struct for jetstream
+type js struct {
+	nc *Conn
+	// For import JS from other accounts.
+	pre string
+}
+
+// Request API subjects for JetStream.
+const (
+	// JSApiAccountInfo is for obtaining general information about JetStream.
+	JSApiAccountInfo = "$JS.API.INFO"
+	// JSApiStreams can lookup a stream by subject.
+	JSApiStreams = "$JS.API.STREAM.NAMES"
+	// JSApiConsumerCreateT is used to create consumers.
+	JSApiConsumerCreateT = "$JS.API.CONSUMER.CREATE.%s"
+	// JSApiDurableCreateT is used to create durable consumers.
+	JSApiDurableCreateT = "$JS.API.CONSUMER.DURABLE.CREATE.%s.%s"
+	// JSApiConsumerInfoT is used to create consumers.
+	JSApiConsumerInfoT = "$JS.API.CONSUMER.INFO.%s.%s"
+	// JSApiRequestNextT is the prefix for the request next message(s) for a consumer in worker/pull mode.
+	JSApiRequestNextT = "$JS.API.CONSUMER.MSG.NEXT.%s.%s"
+	// JSApiStreamCreate is the endpoint to create new streams.
+	JSApiStreamCreateT = "$JS.API.STREAM.CREATE.%s"
+)
+
+// JetStream returns a JetStream context for pub/sub interactions.
+func (nc *Conn) JetStream(optApiPre ...string) (JetStream, error) {
+	js := &js{nc: nc}
+
+	if len(optApiPre) > 0 {
+		const tokSep = "."
+		if len(optApiPre) == 1 && len(optApiPre[0]) > 0 {
+			js.pre = optApiPre[0]
+			if !strings.HasSuffix(js.pre, tokSep) {
+				js.pre = js.pre + tokSep
+			}
+		} else {
+			return nil, ErrJetStreamBadPre
+		}
+	}
+
+	resp, err := nc.Request(js.apiSubj(JSApiAccountInfo), nil, defaultRequestWait)
```

Comment:
should be able to skip this for cases where I dont have the full API only some imported subjects

Reply:
I have this on my list  to think through, essentially attaching to consumers etc. An idea could be for the above JetStream() context options. We could have AttachOnly() as an option that would skip that.

Reply:
ok I gave this one a first pass as a separate change below.